simulation/simulation.py

```diff
@@ -0,0 +1,400 @@
+"""
+Simulate CDM data for testing purposes. Persons are simulated to have hidden disease states. Some states are fixes at
+the start, to simulate fixed traits such as genetics, while the remaining states are dynamic. The probability to enter
+a dynamic disease state depending on the current disease states in a non-linear way (i.e. using interactions between
+states). Concepts are simulated to be observed with probabilities depending on the current disease states, again in a
+non-linear way. Concepts imply visits, and visits in return increase the probability of observing concepts, thus causing
+concepts to cluster in time. The simulated data is saved in the CDM format.
+
+To generate patient-level predicition problems with a gold standard, the model is executed. For each person, an index
+date is randomly selected, and Monte-Carlo simulations are performed to simulate the future.
+"""
+import cProfile
+import configparser
+import json
+import logging
+import multiprocessing
+import os
+import sys
+from typing import List
+
+import numpy as np
+import tqdm as tqdm
+
+import cdm_data
+import utils.logger as logger
+
+LOGGER_FILE_NAME = "_simulation_log.txt"
+QUIET = 0
+CHATTY = 1
+OBSESSIVE = 2
+JSON_FILE_NAME = "_simulation.json"
+PRETRAINING_FOLDER = "pretraining"
+TRAIN_FOLDER = "train"
+TEST_FOLDER = "test"
+
+
+def logistic(x):
+    return 1 / (1 + np.exp(-x))
+
+
+class SimulationSettings:
+    """
+    Settings for the simulation.
+    """
+
+    def __init__(self,
+                 dynamic_state_count: int = 10,
+                 fixed_state_count: int = 5,
+                 concept_count: int = 100,
+                 serious_concept_count: int = 10,
+                 visit_multiplier: int = 2,
+                 days_to_simulate: int = 365 * 2):
+        self.dynamic_state_count = dynamic_state_count
+        self.fixed_state_count = fixed_state_count
+        self.concept_count = concept_count
+        self.serious_concept_count = serious_concept_count
+        self.visit_multiplier = visit_multiplier
+        self.days_to_simulate = days_to_simulate
+
+
+class SimulationTask:
+    partition_count: int
+
+
+class PreTrainingTask(SimulationTask):
+    person_count: int
```

Comment:
Is there a reason you are using a mix of [data class](https://docs.python.org/3/library/dataclasses.html) features and init methods.  Like here you have `person_count: int` under the class definition and `self.person_count = person_count` in the `init` method. The latter is redundant and will be automatically generated from the former. 

In fact you could remove the init methods from `SimulationSettings` and `PreTrainingTask` if you add the variables under the class definition.

I think it would be good to be consistent in using either a data class or init methods where the init methods don't contain any logic, only assignment to class variables.

I haven't used data classes much myself so I could be missing something.

Reply:
Thanks. I'm pretty new to Python, so wasn't aware of the dataclasses module. I'll make sure the data classes only have data.